File: frontend/src/services/documentService.js
import api from './api';

/**
 * Document Service
 * 
 * This service handles all API calls related to documents.
 */
const documentService = {
  /**
   * Get all documents with optional filtering
   * @param {Object} filters - Optional filters
   * @param {string} filters.related_entity_type - Filter by related entity type
   * @param {number} filters.related_entity_id - Filter by related entity ID
   * @param {string} filters.document_type - Filter by document type
   * @param {string} filters.status - Filter by status
   * @param {number} skip - Number of items to skip
   * @param {number} limit - Number of items to return
   * @returns {Promise<Array>} - List of documents
   */
  getDocuments: async (filters = {}, skip = 0, limit = 100) => {
    try {
      let queryParams = `skip=${skip}&limit=${limit}`;
      
      if (filters.related_entity_type) {
        queryParams += `&related_entity_type=${filters.related_entity_type}`;
      }
      
      if (filters.related_entity_id) {
        queryParams += `&related_entity_id=${filters.related_entity_id}`;
      }
      
      if (filters.document_type) {
        queryParams += `&document_type=${filters.document_type}`;
      }
      
      if (filters.status) {
        queryParams += `&status=${filters.status}`;
      }
      
      const response = await api.get(`/documents?${queryParams}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching documents:', error);
      throw error;
    }
  },

  /**
   * Get a document by ID
   * @param {number} id - Document ID
   * @returns {Promise<Object>} - Document
   */
  getDocument: async (id) => {
    try {
      const response = await api.get(`/documents/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching document with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Create a new document
   * @param {Object} documentData - Document data
   * @returns {Promise<Object>} - Created document
   */
  createDocument: async (documentData) => {
    try {
      const response = await api.post('/documents', documentData);
      return response.data;
    } catch (error) {
      console.error('Error creating document:', error);
      throw error;
    }
  },

  /**
   * Upload a new document
   * @param {File} file - File to upload
   * @param {Object} documentData - Document metadata
   * @returns {Promise<Object>} - Created document
   */
  uploadDocument: async (file, documentData) => {
    try {
      const formData = new FormData();
      formData.append('file', file);
      
      // Add document metadata to form data
      Object.keys(documentData).forEach(key => {
        if (documentData[key] !== null && documentData[key] !== undefined) {
          formData.append(key, documentData[key]);
        }
      });
      
      const response = await api.post('/documents/upload', formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });
      
      return response.data;
    } catch (error) {
      console.error('Error uploading document:', error);
      throw error;
    }
  },

  /**
   * Update a document
   * @param {number} id - Document ID
   * @param {Object} documentData - Document data to update
   * @returns {Promise<Object>} - Updated document
   */
  updateDocument: async (id, documentData) => {
    try {
      const response = await api.put(`/documents/${id}`, documentData);
      return response.data;
    } catch (error) {
      console.error(`Error updating document with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Delete a document
   * @param {number} id - Document ID
   * @returns {Promise<boolean>} - Success status
   */
  deleteDocument: async (id) => {
    try {
      const response = await api.delete(`/documents/${id}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting document with ID ${id}:`, error);
      throw error;
    }
  },

  /**
   * Get all versions of a document
   * @param {number} documentId - Document ID
   * @returns {Promise<Array>} - List of document versions
   */
  getDocumentVersions: async (documentId) => {
    try {
      const response = await api.get(`/documents/${documentId}/versions`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching versions for document with ID ${documentId}:`, error);
      throw error;
    }
  },

  /**
   * Get a document version by ID
   * @param {number} versionId - Version ID
   * @returns {Promise<Object>} - Document version
   */
  getDocumentVersion: async (versionId) => {
    try {
      const response = await api.get(`/documents/versions/${versionId}`);
      return response.data;
    } catch (error) {
      console.error(`Error fetching document version with ID ${versionId}:`, error);
      throw error;
    }
  },

  /**
   * Create a new document version
   * @param {number} documentId - Document ID
   * @param {Object} versionData - Version data
   * @returns {Promise<Object>} - Created document version
   */
  createDocumentVersion: async (documentId, versionData) => {
    try {
      const response = await api.post(`/documents/${documentId}/versions`, {
        ...versionData,
        document_id: documentId
      });
      return response.data;
    } catch (error) {
      console.error(`Error creating version for document with ID ${documentId}:`, error);
      throw error;
    }
  },

  /**
   * Upload a new version of a document
   * @param {number} documentId - Document ID
   * @param {File} file - File to upload
   * @param {Object} versionData - Version metadata
   * @returns {Promise<Object>} - Created document version
   */
  uploadDocumentVersion: async (documentId, file, versionData = {}) => {
    try {
      const formData = new FormData();
      formData.append('file', file);
      
      // Add version metadata to form data
      Object.keys(versionData).forEach(key => {
        if (versionData[key] !== null && versionData[key] !== undefined) {
          formData.append(key, versionData[key]);
        }
      });
      
      const response = await api.post(`/documents/${documentId}/upload-version`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });
      
      return response.data;
    } catch (error) {
      console.error(`Error uploading version for document with ID ${documentId}:`, error);
      throw error;
    }
  },

  /**
   * Update a document version
   * @param {number} versionId - Version ID
   * @param {Object} versionData - Version data to update
   * @returns {Promise<Object>} - Updated document version
   */
  updateDocumentVersion: async (versionId, versionData) => {
    try {
      const response = await api.put(`/documents/versions/${versionId}`, versionData);
      return response.data;
    } catch (error) {
      console.error(`Error updating document version with ID ${versionId}:`, error);
      throw error;
    }
  },

  /**
   * Delete a document version
   * @param {number} versionId - Version ID
   * @returns {Promise<boolean>} - Success status
   */
  deleteDocumentVersion: async (versionId) => {
    try {
      const response = await api.delete(`/documents/versions/${versionId}`);
      return response.data;
    } catch (error) {
      console.error(`Error deleting document version with ID ${versionId}:`, error);
      throw error;
    }
  }
};

export default documentService;

File: backend/app/services/document_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from datetime import date
import os
from fastapi import UploadFile, File

from app.models.documents import Document, DocumentVersion
from app.schemas.document import DocumentCreate, DocumentUpdate, DocumentVersionCreate, DocumentVersionUpdate

class DocumentService:
    @staticmethod
    def get_documents(
        db: Session, 
        skip: int = 0, 
        limit: int = 100,
        related_entity_type: Optional[str] = None,
        related_entity_id: Optional[int] = None,
        document_type: Optional[str] = None,
        status: Optional[str] = None
    ) -> List[Document]:
        """Get all documents with optional filtering"""
        query = db.query(Document)
        
        if related_entity_type:
            query = query.filter(Document.related_entity_type == related_entity_type)
        
        if related_entity_id:
            query = query.filter(Document.related_entity_id == related_entity_id)
        
        if document_type:
            query = query.filter(Document.document_type == document_type)
        
        if status:
            query = query.filter(Document.status == status)
        
        return query.offset(skip).limit(limit).all()
    
    @staticmethod
    def get_document(db: Session, document_id: int) -> Optional[Document]:
        """Get a single document by ID"""
        return db.query(Document).filter(Document.id == document_id).first()
    
    @staticmethod
    def create_document(db: Session, document: DocumentCreate) -> Document:
        """Create a new document"""
        db_document = Document(**document.dict())
        db.add(db_document)
        db.commit()
        db.refresh(db_document)
        return db_document
    
    @staticmethod
    async def upload_document(
        db: Session, 
        file: UploadFile, 
        document_data: Dict[str, Any],
        upload_dir: str = "uploads"
    ) -> Document:
        """Upload a new document file and create document record"""
        # Ensure upload directory exists
        os.makedirs(upload_dir, exist_ok=True)
        
        # Generate file path
        file_extension = os.path.splitext(file.filename)[1]
        timestamp = date.today().strftime("%Y%m%d")
        file_name = f"{timestamp}_{file.filename}"
        file_path = os.path.join(upload_dir, file_name)
        
        # Save file
        with open(file_path, "wb") as f:
            content = await file.read()
            f.write(content)
        
        # Create document record
        document_data["file_path"] = file_path
        document_data["name"] = document_data.get("name", file.filename)
        
        document = DocumentCreate(**document_data)
        return DocumentService.create_document(db, document)
    
    @staticmethod
    def update_document(db: Session, document_id: int, document: DocumentUpdate) -> Optional[Document]:
        """Update an existing document"""
        db_document = DocumentService.get_document(db, document_id)
        if db_document:
            update_data = document.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_document, key, value)
            
            db.commit()
            db.refresh(db_document)
        return db_document
    
    @staticmethod
    def delete_document(db: Session, document_id: int) -> bool:
        """Delete a document"""
        db_document = DocumentService.get_document(db, document_id)
        if db_document:
            # Delete the file if it exists
            if os.path.exists(db_document.file_path):
                os.remove(db_document.file_path)
            
            # Delete document versions files
            for version in db_document.versions:
                if os.path.exists(version.file_path):
                    os.remove(version.file_path)
            
            db.delete(db_document)
            db.commit()
            return True
        return False
    
    @staticmethod
    def get_document_versions(db: Session, document_id: int) -> List[DocumentVersion]:
        """Get all versions of a document"""
        return db.query(DocumentVersion).filter(DocumentVersion.document_id == document_id).all()
    
    @staticmethod
    def get_document_version(db: Session, version_id: int) -> Optional[DocumentVersion]:
        """Get a single document version by ID"""
        return db.query(DocumentVersion).filter(DocumentVersion.id == version_id).first()
    
    @staticmethod
    def create_document_version(db: Session, version: DocumentVersionCreate) -> DocumentVersion:
        """Create a new document version"""
        # Get the document to update its latest version
        document = db.query(Document).filter(Document.id == version.document_id).first()
        if document:
            # Update the document's file path to the new version
            document.file_path = version.file_path
            db.commit()
        
        db_version = DocumentVersion(**version.dict())
        db.add(db_version)
        db.commit()
        db.refresh(db_version)
        return db_version
    
    @staticmethod
    async def upload_document_version(
        db: Session, 
        document_id: int,
        file: UploadFile, 
        notes: Optional[str] = None,
        uploaded_by: Optional[int] = None,
        upload_dir: str = "uploads"
    ) -> DocumentVersion:
        """Upload a new version of a document"""
        # Ensure upload directory exists
        os.makedirs(upload_dir, exist_ok=True)
        
        # Get the document
        document = DocumentService.get_document(db, document_id)
        if not document:
            raise ValueError(f"Document with ID {document_id} not found")
        
        # Get the latest version number
        versions = DocumentService.get_document_versions(db, document_id)
        version_number = 1
        if versions:
            version_number = max(v.version_number for v in versions) + 1
        
        # Generate file path
        file_extension = os.path.splitext(file.filename)[1]
        timestamp = date.today().strftime("%Y%m%d")
        file_name = f"{timestamp}_{document.name}_v{version_number}{file_extension}"
        file_path = os.path.join(upload_dir, file_name)
        
        # Save file
        with open(file_path, "wb") as f:
            content = await file.read()
            f.write(content)
        
        # Create version record
        version_data = {
            "document_id": document_id,
            "version_number": version_number,
            "file_path": file_path,
            "uploaded_by": uploaded_by,
            "notes": notes
        }
        
        version = DocumentVersionCreate(**version_data)
        return DocumentService.create_document_version(db, version)
    
    @staticmethod
    def update_document_version(db: Session, version_id: int, version: DocumentVersionUpdate) -> Optional[DocumentVersion]:
        """Update an existing document version"""
        db_version = DocumentService.get_document_version(db, version_id)
        if db_version:
            update_data = version.dict(exclude_unset=True)
            for key, value in update_data.items():
                setattr(db_version, key, value)
            
            db.commit()
            db.refresh(db_version)
        return db_version
    
    @staticmethod
    def delete_document_version(db: Session, version_id: int) -> bool:
        """Delete a document version"""
        db_version = DocumentService.get_document_version(db, version_id)
        if db_version:
            # Delete the file if it exists
            if os.path.exists(db_version.file_path):
                os.remove(db_version.file_path)
            
            db.delete(db_version)
            db.commit()
            return True
        return False
